src/proj.rs

```diff
@@ -0,0 +1,791 @@
+
+    if (det).abs() < 1e-10 {
+        return None;
+    }
+
+    let m10 = t01 * t22 - t02 * t21;
+    let m11 = t00 * t22 - t02 * t20;
+    let m12 = t00 * t21 - t01 * t20;
+    let m20 = t01 * t12 - t02 * t11;
+    let m21 = t00 * t12 - t02 * t10;
+    let m22 = t00 * t11 - t01 * t10;
+
+    let inv = [
+        m00 / det, -m10 / det,  m20 / det,
+       -m01 / det,  m11 / det, -m21 / det,
+        m02 / det, -m12 / det,  m22 / det
+    ];
+
+    Some(normalize(inv))
+}
+
+fn mul3x3(a: [f32; 9], b: [f32; 9]) -> [f32; 9] {
+    let (
+        a11, a12, a13,
+        a21, a22, a23,
+        a31, a32, a33
+        ) = (
+        a[0], a[1], a[2],
+        a[3], a[4], a[5], 
+        a[6], a[7], a[8]
+        );
+    let (
+        b11, b12, b13,
+        b21, b22, b23,
+        b31, b32, b33
+        ) = (
+        b[0], b[1], b[2],
+        b[3], b[4], b[5], 
+        b[6], b[7], b[8]
+        );
+
+    [a11*b11 + a12*b21 + a13*b31, a11*b12 + a12*b22 + a13*b32, a11*b13 + a12*b23 +a13*b33,
+     a21*b11 + a22*b21 + a23*b31, a21*b12 + a22*b22 + a23*b32, a21*b13 + a22*b23 + a23*b33,
+     a31*b11 + a32*b21 + a33*b31, a31*b12 + a32*b22 + a33*b32, a31*b13 + a32*b23 + a33*b33]
+}
+
+fn blend<P>(
+    top_left: P,
+    top_right: P,
+    bottom_left: P,
+    bottom_right: P,
+    right_weight: f32,
+    bottom_weight: f32,
+) -> P
+where
+    P: Pixel,
+    P::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let top = top_left.map2(&top_right, |u, v| {
+        P::Subpixel::clamp((1f32 - right_weight) * cast(u) + right_weight * cast(v))
+    });
+
+    let bottom = bottom_left.map2(&bottom_right, |u, v| {
+        P::Subpixel::clamp((1f32 - right_weight) * cast(u) + right_weight * cast(v))
+    });
+
+    top.map2(&bottom, |u, v| {
+        P::Subpixel::clamp((1f32 - bottom_weight) * cast(u) + bottom_weight * cast(v))
+    })
+}
+
+
+#[allow(deprecated)]
+fn interpolate<P>(image: &Image<P>, x: f32, y: f32, default: P) -> P
+where
+    P: Pixel + 'static,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let left = x.floor();
+    let right = left + 1f32;
+    let top = y.floor();
+    let bottom = top + 1f32;
+
+    let right_weight = x - left;
+    let bottom_weight = y - top;
+
+    // default if out of bound
+    let (width, height) = image.dimensions();
+    if left < 0f32 || right >= width as f32 || top < 0f32 || bottom >= height as f32 {
+        default
+    } else {
+        let (tl, tr, bl, br) = unsafe {
+            (
+                image.unsafe_get_pixel(left as u32, top as u32),
+                image.unsafe_get_pixel(right as u32, top as u32),
+                image.unsafe_get_pixel(left as u32, bottom as u32),
+                image.unsafe_get_pixel(right as u32, bottom as u32),
+            )
+        };
+        blend(tl, tr, bl, br, right_weight, bottom_weight)
+    }
+}
+
+#[inline(always)]
+#[allow(deprecated)]
+fn nearest<P: Pixel + 'static>(image: &Image<P>, x: f32, y: f32, default: P) -> P {
```

Comment:
This one is slightly different to the equivalent function in `affine`, but I'm not sure why.

Reply:
The change was made during performance optimizations, but, to be honest, I am not sure it actually is any faster with this change any more.